### PR TITLE
Add interface and junit implementation

### PR DESCRIPTION
**Pull request by bpsix**

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,32 +1,85 @@
...
     @Override
     public int count() {
-        return 0;
+        //If string is null or blank, return 0.
+        if(this.string == null || this.string.isBlank()) return 0;
+        /*
+        Otherwise, get the number of words in string, which are
+        separated by one or more whitespace characters.
+         */
+        return string.split("[\\s\\r\\t\\n]+").length;
```
Comment (IgorJanGit) [inline src/main/java/StringManipulation.java]: you could for  this just do String[] words = string.split("\\s+");  it less characters and it does job of the rest of  \\r\\t\n. here the definition https://learn.microsoft.com/en-us/dotnet/standard/base-types/character-classes-in-regular-expressions#whitespace-character-s

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,32 +1,85 @@
...
+        //check null string
+        if(this.string == null) return null;
+        //check startWord <= 0
+        if(startWord <= 0) throw new IllegalArgumentException("startWord must be greater than zero");
+        //check endWord <= 0
+        if(endWord <= 0) throw new IllegalArgumentException("endWord must be greater than zero");
+        //check startWord > endWord
+        if(startWord > endWord) throw new IllegalArgumentException("startWord cannot be larger than endWord");
+        //check string.count() < endWord
+        if(this.count() < endWord) throw new IndexOutOfBoundsException("endWord cannot be larger than the total word count");
```
Comment (IgorJanGit) [inline src/main/java/StringManipulation.java]: use to blank likes makes it easier to read the code  also could combine the exception with or statements like this  if (startWord <= 0 || endWord <= 0 || startWord > endWord) {} you could decrease the number of lines used

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,32 +1,85 @@
...
+        //check valid indices
+        for(int i = 0; i < indices.length; i++) {
+            if(indices[i] < 0 || indices[i] >= this.string.length()) throw new IndexOutOfBoundsException("Invalid array index");
+        }
+        //average case
+        StringBuilder b = new StringBuilder();
+        for(int i = 0; i < indices.length; i++) {
+            b.append(this.string.charAt(indices[i]));
+        }
+        return b.toString();
```
Comment (IgorJanGit) [inline src/main/java/StringManipulation.java]: The minor issue about the comment you can expand to explain what is going on with context because comments are redundant with the code it is something I, too, struggle with in my code, the code looks good. i did it slightly differently where can put in IndexOutOfBoundsException in the for loop so it saves space.

**Comment by IgorJanGit** [comment]
Code is good, with tons of functions that I wish I had done in my code, as it is more understandable than my code. A minor issue is that comment is a bit too vague. It says what codes do but not with the context of why it does. A quick explanation of code would be good. I have the same issue with comments in my own code

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,32 +1,85 @@
...
     @Override
     public void setString(String string) {
+        this.string = string;
     }
 
     @Override
     public int count() {
-        return 0;
+        //If string is null or blank, return 0.
+        if(this.string == null || this.string.isBlank()) return 0;
```
Comment (IgorJanGit) [inline src/main/java/StringManipulation.java:20]: This is great i should have done this in my code it fairly simple and works

File: src/main/java/StringManipulation.java [hunk 5]
```diff
@@ -1,32 +1,85 @@
...
-        return null;
+        //check null string
+        if(this.string == null) return null;
+        // check n <= 0
+        if(n <= 0) throw new IllegalArgumentException("N must be greater than 0");
+        // check n > string string
+        if(n > this.string.length()) throw new IndexOutOfBoundsException("Index out of bounds");
+        //edge cases
+        if(n == 1 && !maintainSpacing) return "";
+        if(n == 1 && maintainSpacing) return this.string.replaceAll(".", " ");
```
Comment (IgorJanGit) [inline src/main/java/StringManipulation.java:42]: you could make this as an inbulit feature   of the for loop to save space  and handle edge case

File: src/test/java/StringManipulationTest.java [hunk 6]
```diff
@@ -19,95 +19,193 @@ public void tearDown() {
...
     }
 
+    //Maintain spacing case
     @Test
     public void testRemoveNthCharacter2() {
         manipulatedstring.setString("I'd b3tt3r put s0me d161ts in this 5tr1n6, right?");
         assertEquals("I'  b tt r  ut s0 e  16 ts in th s  tr n6  r gh ?", manipulatedstring.removeNthCharacter(3, true));
     }
 
+    //Throws IndexOutOfBoundsException when n larger than string length
```
Comment (IgorJanGit) [inline src/test/java/StringManipulationTest.java:108]: testing looks good but i would add for exception test  for when it true just to double check if it works

File: src/test/java/StringManipulationTest.java [hunk 7]
```diff
@@ -19,95 +19,193 @@ public void tearDown() {
...
     }
 
+    //n=1, maintain spacing case
     @Test
     public void testRemoveNthCharacter7() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("12345 6789");
+        String result = manipulatedstring.removeNthCharacter(1, true);
+        int length = result.length();
+        assertEquals(10, length);
```
Comment (IgorJanGit) [inline src/test/java/StringManipulationTest.java:147]: did not think about doing this test in my own code looks good

**Comment by bpsix** [comment]
Thanks for the feedback. I went back and added spacing between my comments and removed the redundancies from my regexes.
